<a href="https://colab.research.google.com/github/Wittig18/Sales_data_preprocessing/blob/main/Data_preprocessing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Importing Libraries

import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns

#

In [ ]:
data = pd.read_csv('customer_data.csv')

In [ ]:
data

CustomerID    Age  Gender    Income  PurchaseAmount    JoinDate  \
0          101   25.0    Male   45000.0           120.0   5/10/2021   
1          102    NaN  Female   52000.0           220.0   7/13/2020   
2          103   33.0       M       NaN           310.0  13-08-2022   
3          104  200.0  Female   58000.0             NaN   9/21/2021   
4          105   28.0     NaN   47000.0           150.0   12/2/2021   
5          106   22.0    Male   38000.0            90.0   1/17/2020   
6          107   31.0  Female   62000.0          5000.0  11/11/2019   
7          108   29.0   Fmale   54000.0           180.0   4/25/2021   
8          109   27.0    Male   49000.0           160.0   7/18/2020   
9          110   35.0  Female   61000.0           200.0  2022-02-30   
10         111   26.0    male   45000.0           130.0    5/5/2020   
11         112    NaN  Female   58000.0           300.0   8/19/2021   
12         113   44.0    Male  150000.0           800.0   3/12/2019   
13         114   31.0  Female   60000.0           250.0   6/21/2019   
14         115   29.0  Female   58000.0           260.0   11/3/2021   
15         115   29.0  Female   58000.0           260.0   11/3/2021   
16         116   18.0    Male   22000.0            70.0   1/14/2022   
17         117   21.0     NaN   30000.0         50000.0   7/22/2021   
18         118   29.0    Male       NaN           140.0  12/18/2020   
19         119   32.0  Female   67000.0             NaN    9/5/2021   
20         120   -5.0    Male   50000.0           210.0   1/25/2021   
21         121   27.0  Female   49000.0          8000.0   2/20/2019   
22         122   31.0    Male   51000.0           160.0    3/1/2020   
23         123    NaN  Female   62000.0           260.0   6/16/2020   
24         124   29.0  Female   59000.0           240.0  10/28/2021   
25         125   35.0     NaN   65000.0           450.0   4/12/2021   
26         126   24.0  Female   27000.0           100.0    8/9/2020   
27         127   23.0  Female   28000.0           110.0   3/17/2022   
28         128   40.0    Male   95000.0           350.0   9/25/2019   
29         129   22.0    Male  300000.0           160.0   7/19/2021   
30         130   31.0  female   56000.0           280.0  11/11/2020   
31         131   28.0    Male   51000.0           270.0  13-07-2021   
32         132   26.0  Female   47000.0           150.0   12/4/2021   
33         133   33.0    Male   49000.0             NaN   4/26/2020   
34         134   27.0    male       NaN           190.0   2/21/2021   
35         135   31.0  Female   60000.0             NaN   5/14/2021   
36         136   29.0  Female   61000.0           310.0   1/18/2022   
37         137   28.0   FMale   43000.0           200.0  10/10/2020   
38         138   45.0    Male   70000.0           400.0   9/20/2021   
39         139   46.0  Female   75000.0           420.0   3/22/2021   
40         140   34.0    Male   62000.0           260.0   8/12/2019   
41         141   25.0    Male   48000.0           180.0    6/1/2021   
42         142   29.0  Female   61000.0           290.0         NaN   
43         143   31.0    Male   52000.0           150.0   7/24/2021   
44         144   32.0  Female   58000.0           230.0   5/15/2021   
45         145   22.0    Male   32000.0            80.0   1/30/2020   
46         146  500.0  Female   65000.0           300.0   8/18/2021   
47         147   30.0    Male       NaN           200.0   9/10/2020   
48         148   33.0  Female   53000.0           210.0   2/27/2021   
49         149   31.0     NaN   62000.0           250.0    4/6/2021   
50         150   24.0    Male   41000.0           190.0  12/30/2020   

         Location IsMember  
0           Lagos      Yes  
1           Abuja       No  
2           Lagos      YES  
3    Port harourt       No  
4          Kaduna      Yes  
5           lagos      yes  
6           Abuja       No  
7          Ibadan      Yes  
8          IBADAN       No  
9           Lagos    

In [ ]:
data.head()

CustomerID    Age  Gender   Income  PurchaseAmount    JoinDate  \
0         101   25.0    Male  45000.0           120.0   5/10/2021   
1         102    NaN  Female  52000.0           220.0   7/13/2020   
2         103   33.0       M      NaN           310.0  13-08-2022   
3         104  200.0  Female  58000.0             NaN   9/21/2021   
4         105   28.0     NaN  47000.0           150.0   12/2/2021   

       Location IsMember  
0         Lagos      Yes  
1         Abuja       No  
2         Lagos      YES  
3  Port harourt       No  
4        Kaduna      Yes

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   CustomerID      51 non-null     int64  
 1   Age             48 non-null     float64
 2   Gender          47 non-null     object 
 3   Income          47 non-null     float64
 4   PurchaseAmount  47 non-null     float64
 5   JoinDate        50 non-null     object 
 6   Location        50 non-null     object 
 7   IsMember        51 non-null     object 
dtypes: float64(3), int64(1), object(4)
memory usage: 3.3+ KB


AttributeError: 'NoneType' object has no attribute 'sum'

In [ ]:
data.describe()

CustomerID         Age         Income  PurchaseAmount
count   51.000000   48.000000      47.000000       47.000000
mean   125.294118   42.291667   60553.191489     1558.297872
std     14.505577   72.247199   40522.569401     7339.463705
min    101.000000   -5.000000   22000.000000       70.000000
25%    113.500000   26.000000   47500.000000      160.000000
50%    125.000000   29.000000   56000.000000      220.000000
75%    137.500000   32.250000   61500.000000      295.000000
max    150.000000  500.000000  300000.000000    50000.000000

In [ ]:
data.shape

(51, 8)

In [ ]:
data.duplicated()

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12    False
13    False
14    False
15     True
16    False
17    False
18    False
19    False
20    False
21    False
22    False
23    False
24    False
25    False
26    False
27    False
28    False
29    False
30    False
31    False
32    False
33    False
34    False
35    False
36    False
37    False
38    False
39    False
40    False
41    False
42    False
43    False
44    False
45    False
46    False
47    False
48    False
49    False
50    False
dtype: bool

In [ ]:
df= data.drop_duplicates(keep = 'first')

In [ ]:
df

CustomerID    Age  Gender    Income  PurchaseAmount    JoinDate  \
0          101   25.0    Male   45000.0           120.0   5/10/2021   
1          102    NaN  Female   52000.0           220.0   7/13/2020   
2          103   33.0       M       NaN           310.0  13-08-2022   
3          104  200.0  Female   58000.0             NaN   9/21/2021   
4          105   28.0     NaN   47000.0           150.0   12/2/2021   
5          106   22.0    Male   38000.0            90.0   1/17/2020   
6          107   31.0  Female   62000.0          5000.0  11/11/2019   
7          108   29.0   Fmale   54000.0           180.0   4/25/2021   
8          109   27.0    Male   49000.0           160.0   7/18/2020   
9          110   35.0  Female   61000.0           200.0  2022-02-30   
10         111   26.0    male   45000.0           130.0    5/5/2020   
11         112    NaN  Female   58000.0           300.0   8/19/2021   
12         113   44.0    Male  150000.0           800.0   3/12/2019   
13         114   31.0  Female   60000.0           250.0   6/21/2019   
14         115   29.0  Female   58000.0           260.0   11/3/2021   
16         116   18.0    Male   22000.0            70.0   1/14/2022   
17         117   21.0     NaN   30000.0         50000.0   7/22/2021   
18         118   29.0    Male       NaN           140.0  12/18/2020   
19         119   32.0  Female   67000.0             NaN    9/5/2021   
20         120   -5.0    Male   50000.0           210.0   1/25/2021   
21         121   27.0  Female   49000.0          8000.0   2/20/2019   
22         122   31.0    Male   51000.0           160.0    3/1/2020   
23         123    NaN  Female   62000.0           260.0   6/16/2020   
24         124   29.0  Female   59000.0           240.0  10/28/2021   
25         125   35.0     NaN   65000.0           450.0   4/12/2021   
26         126   24.0  Female   27000.0           100.0    8/9/2020   
27         127   23.0  Female   28000.0           110.0   3/17/2022   
28         128   40.0    Male   95000.0           350.0   9/25/2019   
29         129   22.0    Male  300000.0           160.0   7/19/2021   
30         130   31.0  female   56000.0           280.0  11/11/2020   
31         131   28.0    Male   51000.0           270.0  13-07-2021   
32         132   26.0  Female   47000.0           150.0   12/4/2021   
33         133   33.0    Male   49000.0             NaN   4/26/2020   
34         134   27.0    male       NaN           190.0   2/21/2021   
35         135   31.0  Female   60000.0             NaN   5/14/2021   
36         136   29.0  Female   61000.0           310.0   1/18/2022   
37         137   28.0   FMale   43000.0           200.0  10/10/2020   
38         138   45.0    Male   70000.0           400.0   9/20/2021   
39         139   46.0  Female   75000.0           420.0   3/22/2021   
40         140   34.0    Male   62000.0           260.0   8/12/2019   
41         141   25.0    Male   48000.0           180.0    6/1/2021   
42         142   29.0  Female   61000.0           290.0         NaN   
43         143   31.0    Male   52000.0           150.0   7/24/2021   
44         144   32.0  Female   58000.0           230.0   5/15/2021   
45         145   22.0    Male   32000.0            80.0   1/30/2020   
46         146  500.0  Female   65000.0           300.0   8/18/2021   
47         147   30.0    Male       NaN           200.0   9/10/2020   
48         148   33.0  Female   53000.0           210.0   2/27/2021   
49         149   31.0     NaN   62000.0           250.0    4/6/2021   
50         150   24.0    Male   41000.0           190.0  12/30/2020   

         Location IsMember  
0           Lagos      Yes  
1           Abuja       No  
2           Lagos      YES  
3    Port harourt       No  
4          Kaduna      Yes  
5           lagos      yes  
6           Abuja       No  
7          Ibadan      Yes  
8          IBADAN       No  
9           Lagos      Yes  
10           kano       No  
11         Kaduna      Yes  
12   

In [ ]:
df.shape

(8,)

In [ ]:
df

Index(['CustomerID', 'Age', 'Gender', 'Income', 'PurchaseAmount', 'JoinDate',
       'Location', 'IsMember'],
      dtype='object')

In [ ]:
df = df.columns.str.strip()

In [ ]:
df.head()

AttributeError: 'Index' object has no attribute 'head'

In [ ]:
df.columns = df.columns.str.strip()

In [ ]:
df.head()

CustomerID    Age  Gender   Income  PurchaseAmount    JoinDate  \
0         101   25.0    Male  45000.0           120.0   5/10/2021   
1         102    NaN  Female  52000.0           220.0   7/13/2020   
2         103   33.0       M      NaN           310.0  13-08-2022   
3         104  200.0  Female  58000.0             NaN   9/21/2021   
4         105   28.0     NaN  47000.0           150.0   12/2/2021   

       Location IsMember  
0         Lagos      Yes  
1         Abuja       No  
2         Lagos      YES  
3  Port harourt       No  
4        Kaduna      Yes

In [ ]:
print(df['Gender'].unique())

['Male' 'Female' 'M' nan 'Fmale' 'male' 'female' 'FMale']


In [ ]:
print(df['Location'].unique())



['Lagos' 'Abuja' 'Port harourt' 'Kaduna' 'lagos' 'Ibadan' 'IBADAN' 'kano'
 'Kano' 'Port Harcourt' 'Kadunaa' 'Port-Harcourt' 'kaduna' 'LAgos' nan]


In [ ]:
print(df['IsMember'].unique())

['Yes' 'No' 'YES' 'yes' 'NO']


In [ ]:
df['CustomerID'] = pd.to_numeric(df['CustomerID'], errors='coerce').astype('Int64')

/tmp/ipython-input-1286763177.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['CustomerID'] = pd.to_numeric(df['CustomerID'], errors='coerce').astype('Int64')


In [ ]:
type(df['CustomerID'])

pandas.core.series.Series

In [ ]:
df.sort_values('CustomerID', inplace=True)

/tmp/ipython-input-3968787557.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.sort_values('CustomerID', inplace=True)


In [ ]:
df.head()

CustomerID    Age  Gender   Income  PurchaseAmount    JoinDate  \
0         101   25.0    Male  45000.0           120.0   5/10/2021   
1         102    NaN  Female  52000.0           220.0   7/13/2020   
2         103   33.0       M      NaN           310.0  13-08-2022   
3         104  200.0  Female  58000.0             NaN   9/21/2021   
4         105   28.0     NaN  47000.0           150.0   12/2/2021   

       Location IsMember  
0         Lagos      Yes  
1         Abuja       No  
2         Lagos      YES  
3  Port harourt       No  
4        Kaduna      Yes

In [ ]:

dups = df[df.duplicated(subset=['CustomerID'], keep = False)]
dups

Empty DataFrame
Columns: [CustomerID, Age, Gender, Income, PurchaseAmount, JoinDate, Location, IsMember]
Index: []

In [ ]:
for col in ['Age','PurchaseAmount', 'Income']:
  df[col] = pd.to_numeric(df[col], errors='coerce')

/tmp/ipython-input-3810180587.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[col] = pd.to_numeric(df[col], errors='coerce')


In [ ]:
df[['Age', 'PurchaseAmount', 'Income']].describe()

Age  PurchaseAmount         Income
count   47.000000       46.000000      46.000000
mean    42.574468     1586.521739   60608.695652
std     73.001413     7417.986015   40968.539833
min     -5.000000       70.000000   22000.000000
25%     26.000000      160.000000   47250.000000
50%     29.000000      215.000000   55000.000000
75%     32.500000      297.500000   61750.000000
max    500.000000    50000.000000  300000.000000

In [ ]:
df.loc[(df['Age'] < 0) | (df['Age'] >120), 'Age'] = np.nan

In [ ]:
df.loc[df['Age'] > 120]

Empty DataFrame
Columns: [CustomerID, Age, Gender, Income, PurchaseAmount, JoinDate, Location, IsMember]
Index: []

In [ ]:
df.loc[df['Income'] < 0, 'Income'] = np.nan

In [ ]:
df.loc[df['PurchaseAmount'] < 0, 'PurchaseAmount'] = np.nan

In [ ]:
from numpy._core.numeric import astype
df['Gender'] = df['Gender'].astype(str).str.strip().str.lower()

/tmp/ipython-input-2322556177.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Gender'] = df['Gender'].astype(str).str.strip().str.lower()


In [ ]:
df['IsMember'] = df['IsMember'].astype(str).str.strip().str.lower()

/tmp/ipython-input-372011239.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['IsMember'] = df['IsMember'].astype(str).str.strip().str.lower()


In [ ]:
df['Location'] = df['Location'].astype(str).str.strip().str.lower()

/tmp/ipython-input-1472263998.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Location'] = df['Location'].astype(str).str.strip().str.lower()


In [ ]:
df['Gender']

0       male
1     female
2          m
3     female
4        nan
5       male
6     female
7      fmale
8       male
9     female
10      male
11    female
12      male
13    female
14    female
16      male
17       nan
18      male
19    female
20      male
21    female
22      male
23    female
24    female
25       nan
26    female
27    female
28      male
29      male
30    female
31      male
32    female
33      male
34      male
35    female
36    female
37     fmale
38      male
39    female
40      male
41      male
42    female
43      male
44    female
45      male
46    female
47      male
48    female
49       nan
50      male
Name: Gender, dtype: object

In [ ]:
df['Gender'] = df['Gender'].replace({'male': 'male', 'm': 'male', 'female': 'female', 'f': 'female', 'fmale' : 'female'})


/tmp/ipython-input-2822732941.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Gender'] = df['Gender'].replace({'male': 'male', 'm': 'male', 'female': 'female', 'f': 'female', 'fmale' : 'female'})


In [ ]:
df['Gender'].unique()

array(['male', 'female', 'nan'], dtype=object)

In [ ]:
df["Location"] = df["Location"].replace({
    "lagos":"lagos", "lAgos":"lagos", "lagos ":"lagos", "port harcourt":"port harcourt", "port-harcourt":"port harcourt",
    "port harourt":"port harcourt", "kadunaa":"kaduna", "kano":"kano", "ibadan":"ibadan"
})


df["IsMember"] = df["IsMember"].replace({"yes":"yes","y":"yes","no":"no","n":"no","yes.":"yes","true":"yes","false":"no"})

/tmp/ipython-input-533374438.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["Location"] = df["Location"].replace({
/tmp/ipython-input-533374438.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["IsMember"] = df["IsMember"].replace({"yes":"yes","y":"yes","no":"no","n":"no","yes.":"yes","true":"yes","false":"no"})


In [ ]:
df.head()

CustomerID   Age  Gender   Income  PurchaseAmount    JoinDate  \
0         101  25.0    male  45000.0           120.0   5/10/2021   
1         102   NaN  female  52000.0           220.0   7/13/2020   
2         103  33.0    male      NaN           310.0  13-08-2022   
3         104   NaN  female  58000.0             NaN   9/21/2021   
4         105  28.0     nan  47000.0           150.0   12/2/2021   

        Location IsMember  
0          lagos      yes  
1          abuja       no  
2          lagos      yes  
3  port harcourt       no  
4         kaduna      yes

In [ ]:
df['JoinDate_Parsed'] = pd.to_datetime(df['JoinDate'], errors='coerce')

/tmp/ipython-input-1078335803.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['JoinDate_Parsed'] = pd.to_datetime(df['JoinDate'], errors='coerce')


In [ ]:
incorrect_date = df[df['JoinDate_Parsed'].isna()][['JoinDate']].drop_duplicates()

In [ ]:
incorrect_date

JoinDate
2   13-08-2022
9   2022-02-30
31  13-07-2021
42         NaN

In [ ]:
for col in ["Age","Income","PurchaseAmount"]:
    df[f"{col}_was_missing"] = df[col].isna().astype(int)

/tmp/ipython-input-3731573946.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[f"{col}_was_missing"] = df[col].isna().astype(int)
/tmp/ipython-input-3731573946.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[f"{col}_was_missing"] = df[col].isna().astype(int)
/tmp/ipython-input-3731573946.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.py

In [ ]:
num_imputer = SimpleImputer(strategy="median")
df[["Age","Income","PurchaseAmount"]] = num_imputer.fit_transform(df[["Age","Income","PurchaseAmount"]])


/tmp/ipython-input-261135908.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[["Age","Income","PurchaseAmount"]] = num_imputer.fit_transform(df[["Age","Income","PurchaseAmount"]])


In [ ]:
df_backup

CustomerID   Age  Gender    Income  PurchaseAmount    JoinDate  \
0          101  25.0    male   45000.0           120.0   5/10/2021   
1          102  29.0  female   52000.0           220.0   7/13/2020   
2          103  33.0    male   55000.0           310.0  13-08-2022   
3          104  29.0  female   58000.0           215.0   9/21/2021   
4          105  28.0     nan   47000.0           150.0   12/2/2021   
5          106  22.0    male   38000.0            90.0   1/17/2020   
6          107  31.0  female   62000.0          5000.0  11/11/2019   
7          108  29.0  female   54000.0           180.0   4/25/2021   
8          109  27.0    male   49000.0           160.0   7/18/2020   
9          110  35.0  female   61000.0           200.0  2022-02-30   
10         111  26.0    male   45000.0           130.0    5/5/2020   
11         112  29.0  female   58000.0           300.0   8/19/2021   
12         113  44.0    male  150000.0           800.0   3/12/2019   
13         114  31.0  female   60000.0           250.0   6/21/2019   
14         115  29.0  female   58000.0           260.0   11/3/2021   
16         116  18.0    male   22000.0            70.0   1/14/2022   
17         117  21.0     nan   30000.0         50000.0   7/22/2021   
18         118  29.0    male   55000.0           140.0  12/18/2020   
19         119  32.0  female   67000.0           215.0    9/5/2021   
20         120  29.0    male   50000.0           210.0   1/25/2021   
21         121  27.0  female   49000.0          8000.0   2/20/2019   
22         122  31.0    male   51000.0           160.0    3/1/2020   
23         123  29.0  female   62000.0           260.0   6/16/2020   
24         124  29.0  female   59000.0           240.0  10/28/2021   
25         125  35.0     nan   65000.0           450.0   4/12/2021   
26         126  24.0  female   27000.0           100.0    8/9/2020   
27         127  23.0  female   28000.0           110.0   3/17/2022   
28         128  40.0    male   95000.0           350.0   9/25/2019   
29         129  22.0    male  300000.0           160.0   7/19/2021   
30         130  31.0  female   56000.0           280.0  11/11/2020   
31         131  28.0    male   51000.0           270.0  13-07-2021   
32         132  26.0  female   47000.0           150.0   12/4/2021   
33         133  33.0    male   49000.0           215.0   4/26/2020   
34         134  27.0    male   55000.0           190.0   2/21/2021   
35         135  31.0  female   60000.0           215.0   5/14/2021   
36         136  29.0  female   61000.0           310.0   1/18/2022   
37         137  28.0  female   43000.0           200.0  10/10/2020   
38         138  45.0    male   70000.0           400.0   9/20/2021   
39         139  46.0  female   75000.0           420.0   3/22/2021   
40         140  34.0    male   62000.0           260.0   8/12/2019   
41         141  25.0    male   48000.0           180.0    6/1/2021   
42         142  29.0  female   61000.0           290.0         NaN   
43         143  31.0    male   52000.0           150.0   7/24/2021   
44         144  32.0  female   58000.0           230.0   5/15/2021   
45         145  22.0    male   32000.0            80.0   1/30/2020   
46         146  29.0  female   65000.0           300.0   8/18/2021   
47         147  30.0    male   55000.0           200.0   9/10/2020   
48         148  33.0  female   53000.0           210.0   2/27/2021   
49         149  31.0     nan   62000.0           250.0    4/6/2021   
50         150  24.0    male   41000.0           190.0  12/30/2020   

         Location IsMember JoinDate_Parsed  Age_was_missing  \
0           lagos      yes      2021-05-10                0   
1           abuja       no      2020-07-13                1   
2           lagos      yes             NaT                0   
3   port harcourt       no      2021-09-21                1   
4          kaduna      yes      2021-12-02                0   
5           lagos      yes      2020-01-17             

In [ ]:
 df = df_backup.copy()

In [ ]:
df

CustomerID   Age  Gender    Income  PurchaseAmount    JoinDate  \
0          101  25.0    male   45000.0           120.0   5/10/2021   
1          102  29.0  female   52000.0           220.0   7/13/2020   
2          103  33.0    male   55000.0           310.0  13-08-2022   
3          104  29.0  female   58000.0           215.0   9/21/2021   
4          105  28.0     nan   47000.0           150.0   12/2/2021   
5          106  22.0    male   38000.0            90.0   1/17/2020   
6          107  31.0  female   62000.0          5000.0  11/11/2019   
7          108  29.0  female   54000.0           180.0   4/25/2021   
8          109  27.0    male   49000.0           160.0   7/18/2020   
9          110  35.0  female   61000.0           200.0  2022-02-30   
10         111  26.0    male   45000.0           130.0    5/5/2020   
11         112  29.0  female   58000.0           300.0   8/19/2021   
12         113  44.0    male  150000.0           800.0   3/12/2019   
13         114  31.0  female   60000.0           250.0   6/21/2019   
14         115  29.0  female   58000.0           260.0   11/3/2021   
16         116  18.0    male   22000.0            70.0   1/14/2022   
17         117  21.0     nan   30000.0         50000.0   7/22/2021   
18         118  29.0    male   55000.0           140.0  12/18/2020   
19         119  32.0  female   67000.0           215.0    9/5/2021   
20         120  29.0    male   50000.0           210.0   1/25/2021   
21         121  27.0  female   49000.0          8000.0   2/20/2019   
22         122  31.0    male   51000.0           160.0    3/1/2020   
23         123  29.0  female   62000.0           260.0   6/16/2020   
24         124  29.0  female   59000.0           240.0  10/28/2021   
25         125  35.0     nan   65000.0           450.0   4/12/2021   
26         126  24.0  female   27000.0           100.0    8/9/2020   
27         127  23.0  female   28000.0           110.0   3/17/2022   
28         128  40.0    male   95000.0           350.0   9/25/2019   
29         129  22.0    male  300000.0           160.0   7/19/2021   
30         130  31.0  female   56000.0           280.0  11/11/2020   
31         131  28.0    male   51000.0           270.0  13-07-2021   
32         132  26.0  female   47000.0           150.0   12/4/2021   
33         133  33.0    male   49000.0           215.0   4/26/2020   
34         134  27.0    male   55000.0           190.0   2/21/2021   
35         135  31.0  female   60000.0           215.0   5/14/2021   
36         136  29.0  female   61000.0           310.0   1/18/2022   
37         137  28.0  female   43000.0           200.0  10/10/2020   
38         138  45.0    male   70000.0           400.0   9/20/2021   
39         139  46.0  female   75000.0           420.0   3/22/2021   
40         140  34.0    male   62000.0           260.0   8/12/2019   
41         141  25.0    male   48000.0           180.0    6/1/2021   
42         142  29.0  female   61000.0           290.0         NaN   
43         143  31.0    male   52000.0           150.0   7/24/2021   
44         144  32.0  female   58000.0           230.0   5/15/2021   
45         145  22.0    male   32000.0            80.0   1/30/2020   
46         146  29.0  female   65000.0           300.0   8/18/2021   
47         147  30.0    male   55000.0           200.0   9/10/2020   
48         148  33.0  female   53000.0           210.0   2/27/2021   
49         149  31.0     nan   62000.0           250.0    4/6/2021   
50         150  24.0    male   41000.0           190.0  12/30/2020   

         Location IsMember JoinDate_Parsed  Age_was_missing  \
0           lagos      yes      2021-05-10                0   
1           abuja       no      2020-07-13                1   
2           lagos      yes             NaT                0   
3   port harcourt       no      2021-09-21                1   
4          kaduna      yes      2021-12-02                0   
5           lagos      yes      2020-01-17             

In [ ]:
df['Gender']

0       male
1     female
2       male
3     female
4        nan
5       male
6     female
7     female
8       male
9     female
10      male
11    female
12      male
13    female
14    female
16      male
17       nan
18      male
19    female
20      male
21    female
22      male
23    female
24    female
25       nan
26    female
27    female
28      male
29      male
30    female
31      male
32    female
33      male
34      male
35    female
36    female
37    female
38      male
39    female
40      male
41      male
42    female
43      male
44    female
45      male
46    female
47      male
48    female
49       nan
50      male
Name: Gender, dtype: object

In [ ]:
df['Gender'] = df['Gender'].fillna(df['Gender'].mode()[0])

In [ ]:
df['Gender']= df['Gender'].replace('nan', df['Gender'].mode()[0])
df['Gender']

0       male
1     female
2       male
3     female
4     female
5       male
6     female
7     female
8       male
9     female
10      male
11    female
12      male
13    female
14    female
16      male
17    female
18      male
19    female
20      male
21    female
22      male
23    female
24    female
25    female
26    female
27    female
28      male
29      male
30    female
31      male
32    female
33      male
34      male
35    female
36    female
37    female
38      male
39    female
40      male
41      male
42    female
43      male
44    female
45      male
46    female
47      male
48    female
49    female
50      male
Name: Gender, dtype: object

In [ ]:
df

CustomerID   Age  Gender    Income  PurchaseAmount    JoinDate  \
0          101  25.0    male   45000.0           120.0   5/10/2021   
1          102  29.0  female   52000.0           220.0   7/13/2020   
2          103  33.0    male   55000.0           310.0  13-08-2022   
3          104  29.0  female   58000.0           215.0   9/21/2021   
4          105  28.0  female   47000.0           150.0   12/2/2021   
5          106  22.0    male   38000.0            90.0   1/17/2020   
6          107  31.0  female   62000.0          5000.0  11/11/2019   
7          108  29.0  female   54000.0           180.0   4/25/2021   
8          109  27.0    male   49000.0           160.0   7/18/2020   
9          110  35.0  female   61000.0           200.0  2022-02-30   
10         111  26.0    male   45000.0           130.0    5/5/2020   
11         112  29.0  female   58000.0           300.0   8/19/2021   
12         113  44.0    male  150000.0           800.0   3/12/2019   
13         114  31.0  female   60000.0           250.0   6/21/2019   
14         115  29.0  female   58000.0           260.0   11/3/2021   
16         116  18.0    male   22000.0            70.0   1/14/2022   
17         117  21.0  female   30000.0         50000.0   7/22/2021   
18         118  29.0    male   55000.0           140.0  12/18/2020   
19         119  32.0  female   67000.0           215.0    9/5/2021   
20         120  29.0    male   50000.0           210.0   1/25/2021   
21         121  27.0  female   49000.0          8000.0   2/20/2019   
22         122  31.0    male   51000.0           160.0    3/1/2020   
23         123  29.0  female   62000.0           260.0   6/16/2020   
24         124  29.0  female   59000.0           240.0  10/28/2021   
25         125  35.0  female   65000.0           450.0   4/12/2021   
26         126  24.0  female   27000.0           100.0    8/9/2020   
27         127  23.0  female   28000.0           110.0   3/17/2022   
28         128  40.0    male   95000.0           350.0   9/25/2019   
29         129  22.0    male  300000.0           160.0   7/19/2021   
30         130  31.0  female   56000.0           280.0  11/11/2020   
31         131  28.0    male   51000.0           270.0  13-07-2021   
32         132  26.0  female   47000.0           150.0   12/4/2021   
33         133  33.0    male   49000.0           215.0   4/26/2020   
34         134  27.0    male   55000.0           190.0   2/21/2021   
35         135  31.0  female   60000.0           215.0   5/14/2021   
36         136  29.0  female   61000.0           310.0   1/18/2022   
37         137  28.0  female   43000.0           200.0  10/10/2020   
38         138  45.0    male   70000.0           400.0   9/20/2021   
39         139  46.0  female   75000.0           420.0   3/22/2021   
40         140  34.0    male   62000.0           260.0   8/12/2019   
41         141  25.0    male   48000.0           180.0    6/1/2021   
42         142  29.0  female   61000.0           290.0         NaN   
43         143  31.0    male   52000.0           150.0   7/24/2021   
44         144  32.0  female   58000.0           230.0   5/15/2021   
45         145  22.0    male   32000.0            80.0   1/30/2020   
46         146  29.0  female   65000.0           300.0   8/18/2021   
47         147  30.0    male   55000.0           200.0   9/10/2020   
48         148  33.0  female   53000.0           210.0   2/27/2021   
49         149  31.0  female   62000.0           250.0    4/6/2021   
50         150  24.0    male   41000.0           190.0  12/30/2020   

         Location IsMember JoinDate_Parsed  Age_was_missing  \
0           lagos      yes      2021-05-10                0   
1           abuja       no      2020-07-13                1   
2           lagos      yes             NaT                0   
3   port harcourt       no      2021-09-21                1   
4          kaduna      yes      2021-12-02                0   
5           lagos      yes      2020-01-17             

In [ ]:
gender = df['Gender'] = df['Gender'].fillna(df['Gender'].mode()[0])
#df['Location'] = df['Location'].fillna(df['Location'].mode()[0], inplace = True)
gender

/tmp/ipython-input-2316473549.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  gender = df['Gender'] = df['Gender'].fillna(df['Gender'].mode()[0], inplace = True)
/tmp/ipython-input-2316473549.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  gender = df['Gender'] = df['Gender'].fillna(df['Gender'].mode()[0], inplace = True)
/tmp/ipython-

In [ ]:
df

CustomerID   Age  Gender    Income  PurchaseAmount    JoinDate  \
0          101  25.0    male   45000.0           120.0   5/10/2021   
1          102  29.0  female   52000.0           220.0   7/13/2020   
2          103  33.0    male   55000.0           310.0  13-08-2022   
3          104  29.0  female   58000.0           215.0   9/21/2021   
4          105  28.0  female   47000.0           150.0   12/2/2021   
5          106  22.0    male   38000.0            90.0   1/17/2020   
6          107  31.0  female   62000.0          5000.0  11/11/2019   
7          108  29.0  female   54000.0           180.0   4/25/2021   
8          109  27.0    male   49000.0           160.0   7/18/2020   
9          110  35.0  female   61000.0           200.0  2022-02-30   
10         111  26.0    male   45000.0           130.0    5/5/2020   
11         112  29.0  female   58000.0           300.0   8/19/2021   
12         113  44.0    male  150000.0           800.0   3/12/2019   
13         114  31.0  female   60000.0           250.0   6/21/2019   
14         115  29.0  female   58000.0           260.0   11/3/2021   
16         116  18.0    male   22000.0            70.0   1/14/2022   
17         117  21.0  female   30000.0         50000.0   7/22/2021   
18         118  29.0    male   55000.0           140.0  12/18/2020   
19         119  32.0  female   67000.0           215.0    9/5/2021   
20         120  29.0    male   50000.0           210.0   1/25/2021   
21         121  27.0  female   49000.0          8000.0   2/20/2019   
22         122  31.0    male   51000.0           160.0    3/1/2020   
23         123  29.0  female   62000.0           260.0   6/16/2020   
24         124  29.0  female   59000.0           240.0  10/28/2021   
25         125  35.0  female   65000.0           450.0   4/12/2021   
26         126  24.0  female   27000.0           100.0    8/9/2020   
27         127  23.0  female   28000.0           110.0   3/17/2022   
28         128  40.0    male   95000.0           350.0   9/25/2019   
29         129  22.0    male  300000.0           160.0   7/19/2021   
30         130  31.0  female   56000.0           280.0  11/11/2020   
31         131  28.0    male   51000.0           270.0  13-07-2021   
32         132  26.0  female   47000.0           150.0   12/4/2021   
33         133  33.0    male   49000.0           215.0   4/26/2020   
34         134  27.0    male   55000.0           190.0   2/21/2021   
35         135  31.0  female   60000.0           215.0   5/14/2021   
36         136  29.0  female   61000.0           310.0   1/18/2022   
37         137  28.0  female   43000.0           200.0  10/10/2020   
38         138  45.0    male   70000.0           400.0   9/20/2021   
39         139  46.0  female   75000.0           420.0   3/22/2021   
40         140  34.0    male   62000.0           260.0   8/12/2019   
41         141  25.0    male   48000.0           180.0    6/1/2021   
42         142  29.0  female   61000.0           290.0         NaN   
43         143  31.0    male   52000.0           150.0   7/24/2021   
44         144  32.0  female   58000.0           230.0   5/15/2021   
45         145  22.0    male   32000.0            80.0   1/30/2020   
46         146  29.0  female   65000.0           300.0   8/18/2021   
47         147  30.0    male   55000.0           200.0   9/10/2020   
48         148  33.0  female   53000.0           210.0   2/27/2021   
49         149  31.0  female   62000.0           250.0    4/6/2021   
50         150  24.0    male   41000.0           190.0  12/30/2020   

         Location IsMember JoinDate_Parsed  Age_was_missing  \
0           lagos      yes      2021-05-10                0   
1           abuja       no      2020-07-13                1   
2           lagos      yes             NaT                0   
3   port harcourt       no      2021-09-21                1   
4          kaduna      yes      2021-12-02                0   
5           lagos      yes      2020-01-17             

In [ ]:
df.describe()

CustomerID        Age        Income  PurchaseAmount  \
count        50.0  50.000000      50.00000        50.00000   
mean        125.5  29.600000   60160.00000      1476.80000   
min         101.0  18.000000   22000.00000        70.00000   
25%        113.25  27.000000   48250.00000       160.00000   
50%         125.5  29.000000   55000.00000       215.00000   
75%        137.75  31.000000   61000.00000       287.50000   
max         150.0  46.000000  300000.00000     50000.00000   
std      14.57738   5.620662   39290.83603      7118.69569   

                     JoinDate_Parsed  Age_was_missing  Income_was_missing  \
count                             46        50.000000           50.000000   
mean   2020-12-24 15:39:07.826086912         0.120000            0.080000   
min              2019-02-20 00:00:00         0.000000            0.000000   
25%              2020-06-22 18:00:00         0.000000            0.000000   
50%              2021-03-10 12:00:00         0.000000            0.000000   
75%              2021-08-11 18:00:00         0.000000            0.000000   
max              2022-03-17 00:00:00         1.000000            1.000000   
std                              NaN         0.328261            0.274048   

       PurchaseAmount_was_missing  
count                   50.000000  
mean                     0.080000  
min                      0.000000  
25%                      0.000000  
50%                      0.000000  
75%                      0.000000  
max                      1.000000  
std                      0.274048

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 50 entries, 0 to 50
Data columns (total 12 columns):
 #   Column                      Non-Null Count  Dtype         
---  ------                      --------------  -----         
 0   CustomerID                  50 non-null     Int64         
 1   Age                         50 non-null     float64       
 2   Gender                      50 non-null     object        
 3   Income                      50 non-null     float64       
 4   PurchaseAmount              50 non-null     float64       
 5   JoinDate                    49 non-null     object        
 6   Location                    50 non-null     object        
 7   IsMember                    50 non-null     object        
 8   JoinDate_Parsed             50 non-null     datetime64[ns]
 9   Age_was_missing             50 non-null     int64         
 10  Income_was_missing          50 non-null     int64         
 11  PurchaseAmount_was_missing  50 non-null     int64         
dtypes

In [ ]:
df.drop(columns = 'JoinDate')

CustomerID   Age  Gender    Income  PurchaseAmount       Location  \
0          101  25.0    male   45000.0           120.0          lagos   
1          102  29.0  female   52000.0           220.0          abuja   
2          103  33.0    male   55000.0           310.0          lagos   
3          104  29.0  female   58000.0           215.0  port harcourt   
4          105  28.0  female   47000.0           150.0         kaduna   
5          106  22.0    male   38000.0            90.0          lagos   
6          107  31.0  female   62000.0          5000.0          abuja   
7          108  29.0  female   54000.0           180.0         ibadan   
8          109  27.0    male   49000.0           160.0         ibadan   
9          110  35.0  female   61000.0           200.0          lagos   
10         111  26.0    male   45000.0           130.0           kano   
11         112  29.0  female   58000.0           300.0         kaduna   
12         113  44.0    male  150000.0           800.0          lagos   
13         114  31.0  female   60000.0           250.0          abuja   
14         115  29.0  female   58000.0           260.0          abuja   
16         116  18.0    male   22000.0            70.0           kano   
17         117  21.0  female   30000.0         50000.0          lagos   
18         118  29.0    male   55000.0           140.0  port harcourt   
19         119  32.0  female   67000.0           215.0          abuja   
20         120  29.0    male   50000.0           210.0          lagos   
21         121  27.0  female   49000.0          8000.0         kaduna   
22         122  31.0    male   51000.0           160.0          abuja   
23         123  29.0  female   62000.0           260.0          lagos   
24         124  29.0  female   59000.0           240.0         kaduna   
25         125  35.0  female   65000.0           450.0         ibadan   
26         126  24.0  female   27000.0           100.0          abuja   
27         127  23.0  female   28000.0           110.0           kano   
28         128  40.0    male   95000.0           350.0  port harcourt   
29         129  22.0    male  300000.0           160.0          abuja   
30         130  31.0  female   56000.0           280.0          lagos   
31         131  28.0    male   51000.0           270.0         kaduna   
32         132  26.0  female   47000.0           150.0         kaduna   
33         133  33.0    male   49000.0           215.0          abuja   
34         134  27.0    male   55000.0           190.0          lagos   
35         135  31.0  female   60000.0           215.0            nan   
36         136  29.0  female   61000.0           310.0          abuja   
37         137  28.0  female   43000.0           200.0          abuja   
38         138  45.0    male   70000.0           400.0          lagos   
39         139  46.0  female   75000.0           420.0  port harcourt   
40         140  34.0    male   62000.0           260.0           kano   
41         141  25.0    male   48000.0           180.0         ibadan   
42         142  29.0  female   61000.0           290.0          abuja   
43         143  31.0    male   52000.0           150.0          lagos   
44         144  32.0  female   58000.0           230.0         kaduna   
45         145  22.0    male   32000.0            80.0  port harcourt   
46         146  29.0  female   65000.0           300.0          lagos   
47         147  30.0    male   55000.0           200.0          abuja   
48         148  33.0  female   53000.0           210.0         ibadan   
49         149  31.0  female   62000.0           250.0          lagos   
50         150  24.0    male   41000.0           190.0          abuja   

   IsMember     JoinDate_Parsed  Age_was_missing  Income_was_missing  \
0       yes 2021-05-10 00:00:00                0                   0   
1        no 2020-07-13 00:00:00                1                   0   
2       yes 2021-03-10 12:00:00                0                

In [ ]:
df['JoinDate_Parsed'] = df['JoinDate_Parsed'].fillna(df['JoinDate_Parsed'].median())

In [ ]:
for col in ['Income', 'PurchaseAmount']:
  lower = df[col].quantile(0.05)
  upper = df[col].quantile(0.95)
  df[col] = np.clip(df[col], lower, upper)

In [ ]:
df['Membershipdays'] = (pd.to_datetime('today') - df['JoinDate_Parsed']).dt.days


In [ ]:
df

CustomerID   Age  Gender   Income  PurchaseAmount    JoinDate  \
0          101  25.0    male  45000.0           120.0   5/10/2021   
1          102  29.0  female  52000.0           220.0   7/13/2020   
2          103  33.0    male  55000.0           310.0  13-08-2022   
3          104  29.0  female  58000.0           215.0   9/21/2021   
4          105  28.0  female  47000.0           150.0   12/2/2021   
5          106  22.0    male  38000.0            94.5   1/17/2020   
6          107  31.0  female  62000.0          3110.0  11/11/2019   
7          108  29.0  female  54000.0           180.0   4/25/2021   
8          109  27.0    male  49000.0           160.0   7/18/2020   
9          110  35.0  female  61000.0           200.0  2022-02-30   
10         111  26.0    male  45000.0           130.0    5/5/2020   
11         112  29.0  female  58000.0           300.0   8/19/2021   
12         113  44.0    male  86000.0           800.0   3/12/2019   
13         114  31.0  female  60000.0           250.0   6/21/2019   
14         115  29.0  female  58000.0           260.0   11/3/2021   
16         116  18.0    male  28900.0            94.5   1/14/2022   
17         117  21.0  female  30000.0          3110.0   7/22/2021   
18         118  29.0    male  55000.0           140.0  12/18/2020   
19         119  32.0  female  67000.0           215.0    9/5/2021   
20         120  29.0    male  50000.0           210.0   1/25/2021   
21         121  27.0  female  49000.0          3110.0   2/20/2019   
22         122  31.0    male  51000.0           160.0    3/1/2020   
23         123  29.0  female  62000.0           260.0   6/16/2020   
24         124  29.0  female  59000.0           240.0  10/28/2021   
25         125  35.0  female  65000.0           450.0   4/12/2021   
26         126  24.0  female  28900.0           100.0    8/9/2020   
27         127  23.0  female  28900.0           110.0   3/17/2022   
28         128  40.0    male  86000.0           350.0   9/25/2019   
29         129  22.0    male  86000.0           160.0   7/19/2021   
30         130  31.0  female  56000.0           280.0  11/11/2020   
31         131  28.0    male  51000.0           270.0  13-07-2021   
32         132  26.0  female  47000.0           150.0   12/4/2021   
33         133  33.0    male  49000.0           215.0   4/26/2020   
34         134  27.0    male  55000.0           190.0   2/21/2021   
35         135  31.0  female  60000.0           215.0   5/14/2021   
36         136  29.0  female  61000.0           310.0   1/18/2022   
37         137  28.0  female  43000.0           200.0  10/10/2020   
38         138  45.0    male  70000.0           400.0   9/20/2021   
39         139  46.0  female  75000.0           420.0   3/22/2021   
40         140  34.0    male  62000.0           260.0   8/12/2019   
41         141  25.0    male  48000.0           180.0    6/1/2021   
42         142  29.0  female  61000.0           290.0         NaN   
43         143  31.0    male  52000.0           150.0   7/24/2021   
44         144  32.0  female  58000.0           230.0   5/15/2021   
45         145  22.0    male  32000.0            94.5   1/30/2020   
46         146  29.0  female  65000.0           300.0   8/18/2021   
47         147  30.0    male  55000.0           200.0   9/10/2020   
48         148  33.0  female  53000.0           210.0   2/27/2021   
49         149  31.0  female  62000.0           250.0    4/6/2021   
50         150  24.0    male  41000.0           190.0  12/30/2020   

         Location IsMember     JoinDate_Parsed  Age_was_missing  \
0           lagos      yes 2021-05-10 00:00:00                0   
1           abuja       no 2020-07-13 00:00:00                1   
2           lagos      yes 2021-03-10 12:00:00                0   
3   port harcourt       no 2021-09-21 00:00:00                1   
4          kaduna      yes 2021-12-02 00:00:00                0   
5           lagos      yes 2020-01-17 00:00:00                0   
6           abu

In [ ]:
df

CustomerID   Age  Gender   Income  PurchaseAmount    JoinDate  \
0          101  25.0    male  34460.0           114.1   5/10/2021   
1          102  29.0  female  34460.0           114.1   7/13/2020   
2          103  33.0    male  34460.0           114.1  13-08-2022   
3          104  29.0  female  34460.0           114.1   9/21/2021   
4          105  28.0  female  34460.0           114.1   12/2/2021   
5          106  22.0    male  34460.0            90.0   1/17/2020   
6          107  31.0  female  34460.0           114.1  11/11/2019   
7          108  29.0  female  34460.0           114.1   4/25/2021   
8          109  27.0    male  34460.0           114.1   7/18/2020   
9          110  35.0  female  34460.0           114.1  2022-02-30   
10         111  26.0    male  34460.0           114.1    5/5/2020   
11         112  29.0  female  34460.0           114.1   8/19/2021   
12         113  44.0    male  34460.0           114.1   3/12/2019   
13         114  31.0  female  34460.0           114.1   6/21/2019   
14         115  29.0  female  34460.0           114.1   11/3/2021   
16         116  18.0    male  24450.0            74.9   1/14/2022   
17         117  21.0  female  30000.0           114.1   7/22/2021   
18         118  29.0    male  34460.0           114.1  12/18/2020   
19         119  32.0  female  34460.0           114.1    9/5/2021   
20         120  29.0    male  34460.0           114.1   1/25/2021   
21         121  27.0  female  34460.0           114.1   2/20/2019   
22         122  31.0    male  34460.0           114.1    3/1/2020   
23         123  29.0  female  34460.0           114.1   6/16/2020   
24         124  29.0  female  34460.0           114.1  10/28/2021   
25         125  35.0  female  34460.0           114.1   4/12/2021   
26         126  24.0  female  27000.0           100.0    8/9/2020   
27         127  23.0  female  28000.0           110.0   3/17/2022   
28         128  40.0    male  34460.0           114.1   9/25/2019   
29         129  22.0    male  34460.0           114.1   7/19/2021   
30         130  31.0  female  34460.0           114.1  11/11/2020   
31         131  28.0    male  34460.0           114.1  13-07-2021   
32         132  26.0  female  34460.0           114.1   12/4/2021   
33         133  33.0    male  34460.0           114.1   4/26/2020   
34         134  27.0    male  34460.0           114.1   2/21/2021   
35         135  31.0  female  34460.0           114.1   5/14/2021   
36         136  29.0  female  34460.0           114.1   1/18/2022   
37         137  28.0  female  34460.0           114.1  10/10/2020   
38         138  45.0    male  34460.0           114.1   9/20/2021   
39         139  46.0  female  34460.0           114.1   3/22/2021   
40         140  34.0    male  34460.0           114.1   8/12/2019   
41         141  25.0    male  34460.0           114.1    6/1/2021   
42         142  29.0  female  34460.0           114.1         NaN   
43         143  31.0    male  34460.0           114.1   7/24/2021   
44         144  32.0  female  34460.0           114.1   5/15/2021   
45         145  22.0    male  32000.0            80.0   1/30/2020   
46         146  29.0  female  34460.0           114.1   8/18/2021   
47         147  30.0    male  34460.0           114.1   9/10/2020   
48         148  33.0  female  34460.0           114.1   2/27/2021   
49         149  31.0  female  34460.0           114.1    4/6/2021   
50         150  24.0    male  34460.0           114.1  12/30/2020   

         Location IsMember     JoinDate_Parsed  Age_was_missing  \
0           lagos      yes 2021-05-10 00:00:00                0   
1           abuja       no 2020-07-13 00:00:00                1   
2           lagos      yes 2021-03-10 12:00:00                0   
3   port harcourt       no 2021-09-21 00:00:00                1   
4          kaduna      yes 2021-12-02 00:00:00                0   
5           lagos      yes 2020-01-17 00:00:00                0   
6           abu

In [ ]:
df['Purchase_bin'] = pd.cut(df['PurchaseAmount'], bins = [0,100,200,500,1000, np.inf], labels = ['very_low', 'low', 'mid', 'high', 'very_high'])

In [ ]:
df

CustomerID   Age  Gender   Income  PurchaseAmount    JoinDate  \
0          101  25.0    male  45000.0           120.0   5/10/2021   
1          102  29.0  female  52000.0           220.0   7/13/2020   
2          103  33.0    male  55000.0           310.0  13-08-2022   
3          104  29.0  female  58000.0           215.0   9/21/2021   
4          105  28.0  female  47000.0           150.0   12/2/2021   
5          106  22.0    male  38000.0            94.5   1/17/2020   
6          107  31.0  female  62000.0          3110.0  11/11/2019   
7          108  29.0  female  54000.0           180.0   4/25/2021   
8          109  27.0    male  49000.0           160.0   7/18/2020   
9          110  35.0  female  61000.0           200.0  2022-02-30   
10         111  26.0    male  45000.0           130.0    5/5/2020   
11         112  29.0  female  58000.0           300.0   8/19/2021   
12         113  44.0    male  86000.0           800.0   3/12/2019   
13         114  31.0  female  60000.0           250.0   6/21/2019   
14         115  29.0  female  58000.0           260.0   11/3/2021   
16         116  18.0    male  28900.0            94.5   1/14/2022   
17         117  21.0  female  30000.0          3110.0   7/22/2021   
18         118  29.0    male  55000.0           140.0  12/18/2020   
19         119  32.0  female  67000.0           215.0    9/5/2021   
20         120  29.0    male  50000.0           210.0   1/25/2021   
21         121  27.0  female  49000.0          3110.0   2/20/2019   
22         122  31.0    male  51000.0           160.0    3/1/2020   
23         123  29.0  female  62000.0           260.0   6/16/2020   
24         124  29.0  female  59000.0           240.0  10/28/2021   
25         125  35.0  female  65000.0           450.0   4/12/2021   
26         126  24.0  female  28900.0           100.0    8/9/2020   
27         127  23.0  female  28900.0           110.0   3/17/2022   
28         128  40.0    male  86000.0           350.0   9/25/2019   
29         129  22.0    male  86000.0           160.0   7/19/2021   
30         130  31.0  female  56000.0           280.0  11/11/2020   
31         131  28.0    male  51000.0           270.0  13-07-2021   
32         132  26.0  female  47000.0           150.0   12/4/2021   
33         133  33.0    male  49000.0           215.0   4/26/2020   
34         134  27.0    male  55000.0           190.0   2/21/2021   
35         135  31.0  female  60000.0           215.0   5/14/2021   
36         136  29.0  female  61000.0           310.0   1/18/2022   
37         137  28.0  female  43000.0           200.0  10/10/2020   
38         138  45.0    male  70000.0           400.0   9/20/2021   
39         139  46.0  female  75000.0           420.0   3/22/2021   
40         140  34.0    male  62000.0           260.0   8/12/2019   
41         141  25.0    male  48000.0           180.0    6/1/2021   
42         142  29.0  female  61000.0           290.0         NaN   
43         143  31.0    male  52000.0           150.0   7/24/2021   
44         144  32.0  female  58000.0           230.0   5/15/2021   
45         145  22.0    male  32000.0            94.5   1/30/2020   
46         146  29.0  female  65000.0           300.0   8/18/2021   
47         147  30.0    male  55000.0           200.0   9/10/2020   
48         148  33.0  female  53000.0           210.0   2/27/2021   
49         149  31.0  female  62000.0           250.0    4/6/2021   
50         150  24.0    male  41000.0           190.0  12/30/2020   

         Location IsMember     JoinDate_Parsed  Age_was_missing  \
0           lagos      yes 2021-05-10 00:00:00                0   
1           abuja       no 2020-07-13 00:00:00                1   
2           lagos      yes 2021-03-10 12:00:00                0   
3   port harcourt       no 2021-09-21 00:00:00                1   
4          kaduna      yes 2021-12-02 00:00:00                0   
5           lagos      yes 2020-01-17 00:00:00                0   
6           abu

In [ ]:
df_ready = pd.get_dummies(df, columns = ['Gender', 'Location', 'IsMember', 'Purchase_bin'], drop_first= True)

In [ ]:
df_ready.drop(columns= 'Location_nan')

CustomerID   Age   Income  PurchaseAmount    JoinDate     JoinDate_Parsed  \
0          101  25.0  45000.0           120.0   5/10/2021 2021-05-10 00:00:00   
1          102  29.0  52000.0           220.0   7/13/2020 2020-07-13 00:00:00   
2          103  33.0  55000.0           310.0  13-08-2022 2021-03-10 12:00:00   
3          104  29.0  58000.0           215.0   9/21/2021 2021-09-21 00:00:00   
4          105  28.0  47000.0           150.0   12/2/2021 2021-12-02 00:00:00   
5          106  22.0  38000.0            94.5   1/17/2020 2020-01-17 00:00:00   
6          107  31.0  62000.0          3110.0  11/11/2019 2019-11-11 00:00:00   
7          108  29.0  54000.0           180.0   4/25/2021 2021-04-25 00:00:00   
8          109  27.0  49000.0           160.0   7/18/2020 2020-07-18 00:00:00   
9          110  35.0  61000.0           200.0  2022-02-30 2021-03-10 12:00:00   
10         111  26.0  45000.0           130.0    5/5/2020 2020-05-05 00:00:00   
11         112  29.0  58000.0           300.0   8/19/2021 2021-08-19 00:00:00   
12         113  44.0  86000.0           800.0   3/12/2019 2019-03-12 00:00:00   
13         114  31.0  60000.0           250.0   6/21/2019 2019-06-21 00:00:00   
14         115  29.0  58000.0           260.0   11/3/2021 2021-11-03 00:00:00   
16         116  18.0  28900.0            94.5   1/14/2022 2022-01-14 00:00:00   
17         117  21.0  30000.0          3110.0   7/22/2021 2021-07-22 00:00:00   
18         118  29.0  55000.0           140.0  12/18/2020 2020-12-18 00:00:00   
19         119  32.0  67000.0           215.0    9/5/2021 2021-09-05 00:00:00   
20         120  29.0  50000.0           210.0   1/25/2021 2021-01-25 00:00:00   
21         121  27.0  49000.0          3110.0   2/20/2019 2019-02-20 00:00:00   
22         122  31.0  51000.0           160.0    3/1/2020 2020-03-01 00:00:00   
23         123  29.0  62000.0           260.0   6/16/2020 2020-06-16 00:00:00   
24         124  29.0  59000.0           240.0  10/28/2021 2021-10-28 00:00:00   
25         125  35.0  65000.0           450.0   4/12/2021 2021-04-12 00:00:00   
26         126  24.0  28900.0           100.0    8/9/2020 2020-08-09 00:00:00   
27         127  23.0  28900.0           110.0   3/17/2022 2022-03-17 00:00:00   
28         128  40.0  86000.0           350.0   9/25/2019 2019-09-25 00:00:00   
29         129  22.0  86000.0           160.0   7/19/2021 2021-07-19 00:00:00   
30         130  31.0  56000.0           280.0  11/11/2020 2020-11-11 00:00:00   
31         131  28.0  51000.0           270.0  13-07-2021 2021-03-10 12:00:00   
32         132  26.0  47000.0           150.0   12/4/2021 2021-12-04 00:00:00   
33         133  33.0  49000.0           215.0   4/26/2020 2020-04-26 00:00:00   
34         134  27.0  55000.0           190.0   2/21/2021 2021-02-21 00:00:00   
35         135  31.0  60000.0           215.0   5/14/2021 2021-05-14 00:00:00   
36         136  29.0  61000.0           310.0   1/18/2022 2022-01-18 00:00:00   
37         137  28.0  43000.0           200.0  10/10/2020 2020-10-10 00:00:00   
38         138  45.0  70000.0           400.0   9/20/2021 2021-09-20 00:00:00   
39         139  46.0  75000.0           420.0   3/22/2021 2021-03-22 00:00:00   
40         140  34.0  62000.0           260.0   8/12/2019 2019-08-12 00:00:00   
41         141  25.0  48000.0           180.0    6/1/2021 2021-06-01 00:00:00   
42         142  29.0  61000.0           290.0         NaN 2021-03-10 12:00:00   
43         143  31.0  52000.0           150.0   7/24/2021 2021-07-24 00:00:00   
44         144  32.0  58000.0           230.0   5/15/2021 2021-05-15 00:00:00   
45         145  22.0  32000.0            94.5   1/30/2020 2020-01-30 00:00:00   
46         146  29.0  65000.0           300.0   8/18/2021 2021-08-18 00:00:00   
47         147  30.0  55000.0           200.0   9/10/2020 2020-09-10 00:00:00   
48         148  33.0  53000.0           210.0   2/27/2021 2021-02-27 00:00:00   
49         149  31.0  62000.0      

In [ ]:
df_ready.columns

Index(['CustomerID', 'Age', 'Income', 'PurchaseAmount', 'JoinDate',
       'JoinDate_Parsed', 'Age_was_missing', 'Income_was_missing',
       'PurchaseAmount_was_missing', 'Membershipdays', 'Gender_male',
       'Location_ibadan', 'Location_kaduna', 'Location_kano', 'Location_lagos',
       'Location_nan', 'Location_port harcourt', 'IsMember_yes',
       'Purchase_bin_low', 'Purchase_bin_mid', 'Purchase_bin_high',
       'Purchase_bin_very_high'],
      dtype='object')

In [ ]:
x = df_ready.drop(columns = ['Income', 'JoinDate_Parsed', 'JoinDate', 'CustomerID'])

In [ ]:
y = df_ready['Income']

In [ ]:
x

Age  PurchaseAmount  Age_was_missing  Income_was_missing  \
0   25.0           120.0                0                   0   
1   29.0           220.0                1                   0   
2   33.0           310.0                0                   1   
3   29.0           215.0                1                   0   
4   28.0           150.0                0                   0   
5   22.0            94.5                0                   0   
6   31.0          3110.0                0                   0   
7   29.0           180.0                0                   0   
8   27.0           160.0                0                   0   
9   35.0           200.0                0                   0   
10  26.0           130.0                0                   0   
11  29.0           300.0                1                   0   
12  44.0           800.0                0                   0   
13  31.0           250.0                0                   0   
14  29.0           260.0                0                   0   
16  18.0            94.5                0                   0   
17  21.0          3110.0                0                   0   
18  29.0           140.0                0                   1   
19  32.0           215.0                0                   0   
20  29.0           210.0                1                   0   
21  27.0          3110.0                0                   0   
22  31.0           160.0                0                   0   
23  29.0           260.0                1                   0   
24  29.0           240.0                0                   0   
25  35.0           450.0                0                   0   
26  24.0           100.0                0                   0   
27  23.0           110.0                0                   0   
28  40.0           350.0                0                   0   
29  22.0           160.0                0                   0   
30  31.0           280.0                0                   0   
31  28.0           270.0                0                   0   
32  26.0           150.0                0                   0   
33  33.0           215.0                0                   0   
34  27.0           190.0                0                   1   
35  31.0           215.0                0                   0   
36  29.0           310.0                0                   0   
37  28.0           200.0                0                   0   
38  45.0           400.0                0                   0   
39  46.0           420.0                0                   0   
40  34.0           260.0                0                   0   
41  25.0           180.0                0                   0   
42  29.0           290.0                0                   0   
43  31.0           150.0                0                   0   
44  32.0           230.0                0                   0   
45  22.0            94.5                0                   0   
46  29.0           300.0                1                   0   
47  30.0           200.0                0                   1   
48  33.0           210.0                0                   0   
49  31.0           250.0                0                   0   
50  24.0           190.0                0                   0   

    PurchaseAmount_was_missing  Membershipdays  Gender_male  Location_ibadan  \
0                            0            1662         True            False   
1                            0            1963        False            False   
2                            0            1723         True            False   
3                            1            1528        False            False   
4                            0            1456        False            False   
5                            0            2141         True            False   
6                            0            2208        False            False   
7                            0            1677   

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.metrics import mean_squared_error, r2_score

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(x,y, test_size= 0.2, random_state= 42)

In [ ]:
x_test

Age  PurchaseAmount  Age_was_missing  Income_was_missing  \
13  31.0           250.0                0                   0   
40  34.0           260.0                0                   0   
31  28.0           270.0                0                   0   
46  29.0           300.0                1                   0   
18  29.0           140.0                0                   1   
49  31.0           250.0                0                   0   
27  23.0           110.0                0                   0   
26  24.0           100.0                0                   0   
33  33.0           215.0                0                   0   
20  29.0           210.0                1                   0   

    PurchaseAmount_was_missing  Membershipdays  Gender_male  Location_ibadan  \
13                           0            2351        False            False   
40                           0            2299         True            False   
31                           0            1723         True            False   
46                           0            1562        False            False   
18                           0            1805         True            False   
49                           0            1696        False            False   
27                           0            1351        False            False   
26                           0            1936        False            False   
33                           1            2041         True            False   
20                           0            1767         True            False   

    Location_kaduna  Location_kano  Location_lagos  Location_nan  \
13            False          False           False         False   
40            False           True           False         False   
31             True          False           False         False   
46            False          False            True         False   
18            False          False           False         False   
49            False          False            True         False   
27            False           True           False         False   
26            False          False           False         False   
33            False          False           False         False   
20            False          False            True         False   

    Location_port harcourt  IsMember_yes  Purchase_bin_low  Purchase_bin_mid  \
13                   False         False             False              True   
40                   False          True             False              True   
31                   False          True             False              True   
46                   False         False             False              True   
18                    True          True              True             False   
49                   False         False             False              True   
27                   False         False              True             False   
26                   False          True             False             False   
33                   False          True             False              True   
20                   False          True             False              True   

    Purchase_bin_high  Purchase_bin_very_high  
13              False                   False  
40              False                   False  
31              False                   False  
46              False                   False  
18              False                   False  
49              False                   False  
27              False                   False  
26              False                   False  
33              False                   False  
20              False                   False

In [ ]:
model = LinearRegression()

In [ ]:
model_2 = RandomForestRegressor(
    n_estimators=100,
    max_depth=None,
    min_samples_split=2,
    min_samples_leaf=1
)

In [ ]:
model_3 = GradientBoostingRegressor()

In [ ]:
model.fit(x_train, y_train)

LinearRegression()

In [ ]:
model_2.fit(x_train, y_train)

RandomForestRegressor()

In [ ]:
model_3.fit(x_train, y_train)

GradientBoostingRegressor()

In [ ]:
y_pred = model.predict(x_test)

In [ ]:
y_pred

array([60231.04494621, 64988.70879381, 61560.07597636, 53651.60784535,
       54775.61469477, 53557.82031551, 38727.86605886, 34545.50832846,
       70907.88730372, 59738.98015453])

In [ ]:
y_pred_2 = model_2.predict(x_test)

In [ ]:
y_pred_3 = model_3.predict(x_test)

In [ ]:
r2_modelRFG = r2_score(y_test, y_pred_2)
r2_modelRFG

0.4662928243793587

In [ ]:
r2_modelGBR = r2_score(y_test, y_pred_3)
r2_modelGBR

0.4957854725120613

In [ ]:
r2_modelLR = r2_score(y_test, y_pred)

In [ ]:
r2_modelLR

0.3248451075015393

In [ ]:
df_ready.drop(columns= 'JoinDate')


CustomerID   Age   Income  PurchaseAmount     JoinDate_Parsed  \
0          101  25.0  45000.0           120.0 2021-05-10 00:00:00   
1          102  29.0  52000.0           220.0 2020-07-13 00:00:00   
2          103  33.0  55000.0           310.0 2021-03-10 12:00:00   
3          104  29.0  58000.0           215.0 2021-09-21 00:00:00   
4          105  28.0  47000.0           150.0 2021-12-02 00:00:00   
5          106  22.0  38000.0            94.5 2020-01-17 00:00:00   
6          107  31.0  62000.0          3110.0 2019-11-11 00:00:00   
7          108  29.0  54000.0           180.0 2021-04-25 00:00:00   
8          109  27.0  49000.0           160.0 2020-07-18 00:00:00   
9          110  35.0  61000.0           200.0 2021-03-10 12:00:00   
10         111  26.0  45000.0           130.0 2020-05-05 00:00:00   
11         112  29.0  58000.0           300.0 2021-08-19 00:00:00   
12         113  44.0  86000.0           800.0 2019-03-12 00:00:00   
13         114  31.0  60000.0           250.0 2019-06-21 00:00:00   
14         115  29.0  58000.0           260.0 2021-11-03 00:00:00   
16         116  18.0  28900.0            94.5 2022-01-14 00:00:00   
17         117  21.0  30000.0          3110.0 2021-07-22 00:00:00   
18         118  29.0  55000.0           140.0 2020-12-18 00:00:00   
19         119  32.0  67000.0           215.0 2021-09-05 00:00:00   
20         120  29.0  50000.0           210.0 2021-01-25 00:00:00   
21         121  27.0  49000.0          3110.0 2019-02-20 00:00:00   
22         122  31.0  51000.0           160.0 2020-03-01 00:00:00   
23         123  29.0  62000.0           260.0 2020-06-16 00:00:00   
24         124  29.0  59000.0           240.0 2021-10-28 00:00:00   
25         125  35.0  65000.0           450.0 2021-04-12 00:00:00   
26         126  24.0  28900.0           100.0 2020-08-09 00:00:00   
27         127  23.0  28900.0           110.0 2022-03-17 00:00:00   
28         128  40.0  86000.0           350.0 2019-09-25 00:00:00   
29         129  22.0  86000.0           160.0 2021-07-19 00:00:00   
30         130  31.0  56000.0           280.0 2020-11-11 00:00:00   
31         131  28.0  51000.0           270.0 2021-03-10 12:00:00   
32         132  26.0  47000.0           150.0 2021-12-04 00:00:00   
33         133  33.0  49000.0           215.0 2020-04-26 00:00:00   
34         134  27.0  55000.0           190.0 2021-02-21 00:00:00   
35         135  31.0  60000.0           215.0 2021-05-14 00:00:00   
36         136  29.0  61000.0           310.0 2022-01-18 00:00:00   
37         137  28.0  43000.0           200.0 2020-10-10 00:00:00   
38         138  45.0  70000.0           400.0 2021-09-20 00:00:00   
39         139  46.0  75000.0           420.0 2021-03-22 00:00:00   
40         140  34.0  62000.0           260.0 2019-08-12 00:00:00   
41         141  25.0  48000.0           180.0 2021-06-01 00:00:00   
42         142  29.0  61000.0           290.0 2021-03-10 12:00:00   
43         143  31.0  52000.0           150.0 2021-07-24 00:00:00   
44         144  32.0  58000.0           230.0 2021-05-15 00:00:00   
45         145  22.0  32000.0            94.5 2020-01-30 00:00:00   
46         146  29.0  65000.0           300.0 2021-08-18 00:00:00   
47         147  30.0  55000.0           200.0 2020-09-10 00:00:00   
48         148  33.0  53000.0           210.0 2021-02-27 00:00:00   
49         149  31.0  62000.0           250.0 2021-04-06 00:00:00   
50         150  24.0  41000.0           190.0 2020-12-30 00:00:00   

    Age_was_missing  Income_was_missing  PurchaseAmount_was_missing  \
0                 0                   0                           0   
1                 1                   0                           0   
2                 0                   1                           0   
3                 1                   0                           1   
4                 0                   0                           0   
5                 0                   0                   

In [ ]:
df['IsMember'].isna().sum()

np.int64(0)

In [ ]:
df['JoinDate_Parsed'] = df['JoinDate_Parsed'].fillna(df['JoinDate_Parsed'].median())

In [ ]:
df

CustomerID   Age  Gender    Income  PurchaseAmount    JoinDate  \
0          101  25.0    male   45000.0           120.0   5/10/2021   
1          102  29.0  female   52000.0           220.0   7/13/2020   
2          103  33.0    male   55000.0           310.0  13-08-2022   
3          104  29.0  female   58000.0           215.0   9/21/2021   
4          105  28.0  female   47000.0           150.0   12/2/2021   
5          106  22.0    male   38000.0            90.0   1/17/2020   
6          107  31.0  female   62000.0          5000.0  11/11/2019   
7          108  29.0  female   54000.0           180.0   4/25/2021   
8          109  27.0    male   49000.0           160.0   7/18/2020   
9          110  35.0  female   61000.0           200.0  2022-02-30   
10         111  26.0    male   45000.0           130.0    5/5/2020   
11         112  29.0  female   58000.0           300.0   8/19/2021   
12         113  44.0    male  150000.0           800.0   3/12/2019   
13         114  31.0  female   60000.0           250.0   6/21/2019   
14         115  29.0  female   58000.0           260.0   11/3/2021   
16         116  18.0    male   22000.0            70.0   1/14/2022   
17         117  21.0  female   30000.0         50000.0   7/22/2021   
18         118  29.0    male   55000.0           140.0  12/18/2020   
19         119  32.0  female   67000.0           215.0    9/5/2021   
20         120  29.0    male   50000.0           210.0   1/25/2021   
21         121  27.0  female   49000.0          8000.0   2/20/2019   
22         122  31.0    male   51000.0           160.0    3/1/2020   
23         123  29.0  female   62000.0           260.0   6/16/2020   
24         124  29.0  female   59000.0           240.0  10/28/2021   
25         125  35.0  female   65000.0           450.0   4/12/2021   
26         126  24.0  female   27000.0           100.0    8/9/2020   
27         127  23.0  female   28000.0           110.0   3/17/2022   
28         128  40.0    male   95000.0           350.0   9/25/2019   
29         129  22.0    male  300000.0           160.0   7/19/2021   
30         130  31.0  female   56000.0           280.0  11/11/2020   
31         131  28.0    male   51000.0           270.0  13-07-2021   
32         132  26.0  female   47000.0           150.0   12/4/2021   
33         133  33.0    male   49000.0           215.0   4/26/2020   
34         134  27.0    male   55000.0           190.0   2/21/2021   
35         135  31.0  female   60000.0           215.0   5/14/2021   
36         136  29.0  female   61000.0           310.0   1/18/2022   
37         137  28.0  female   43000.0           200.0  10/10/2020   
38         138  45.0    male   70000.0           400.0   9/20/2021   
39         139  46.0  female   75000.0           420.0   3/22/2021   
40         140  34.0    male   62000.0           260.0   8/12/2019   
41         141  25.0    male   48000.0           180.0    6/1/2021   
42         142  29.0  female   61000.0           290.0         NaN   
43         143  31.0    male   52000.0           150.0   7/24/2021   
44         144  32.0  female   58000.0           230.0   5/15/2021   
45         145  22.0    male   32000.0            80.0   1/30/2020   
46         146  29.0  female   65000.0           300.0   8/18/2021   
47         147  30.0    male   55000.0           200.0   9/10/2020   
48         148  33.0  female   53000.0           210.0   2/27/2021   
49         149  31.0  female   62000.0           250.0    4/6/2021   
50         150  24.0    male   41000.0           190.0  12/30/2020   

         Location IsMember     JoinDate_Parsed  Age_was_missing  \
0           lagos      yes 2021-05-10 00:00:00                0   
1           abuja       no 2020-07-13 00:00:00                1   
2           lagos      yes 2021-03-10 12:00:00                0   
3   port harcourt       no 2021-09-21 00:00:00                1   
4          kaduna      yes 2021-12-02 00:00:00                0   
5           lagos      yes 2020

In [ ]:
x.head

<bound method NDFrame.head of     CustomerID    Age  Gender  PurchaseAmount    JoinDate       Location  \
0          101   25.0    Male           120.0   5/10/2021          Lagos   
1          102    NaN  Female           220.0   7/13/2020          Abuja   
2          103   33.0       M           310.0  13-08-2022          Lagos   
3          104  200.0  Female             NaN   9/21/2021   Port harourt   
4          105   28.0     NaN           150.0   12/2/2021         Kaduna   
5          106   22.0    Male            90.0   1/17/2020          lagos   
6          107   31.0  Female          5000.0  11/11/2019          Abuja   
7          108   29.0   Fmale           180.0   4/25/2021         Ibadan   
8          109   27.0    Male           160.0   7/18/2020         IBADAN   
9          110   35.0  Female           200.0  2022-02-30          Lagos   
10         111   26.0    male           130.0    5/5/2020           kano   
11         112    NaN  Female           300.0   8/19/2021         Kaduna   
12         113   44.0    Male           800.0   3/12/2019          Lagos   
13         114   31.0  Female           250.0   6/21/2019          Abuja   
14         115   29.0  Female           260.0   11/3/2021          Abuja   
15         115   29.0  Female           260.0   11/3/2021          Abuja   
16         116   18.0    Male            70.0   1/14/2022           Kano   
17         117   21.0     NaN         50000.0   7/22/2021          Lagos   
18         118   29.0    Male           140.0  12/18/2020  Port Harcourt   
19         119   32.0  Female             NaN    9/5/2021          Abuja   
20         120   -5.0    Male           210.0   1/25/2021          Lagos   
21         121   27.0  Female          8000.0   2/20/2019        Kadunaa   
22         122   31.0    Male           160.0    3/1/2020          Abuja   
23         123    NaN  Female           260.0   6/16/2020          Lagos   
24         124   29.0  Female           240.0  10/28/2021         Kaduna   
25         125   35.0     NaN           450.0   4/12/2021         Ibadan   
26         126   24.0  Female           100.0    8/9/2020          Abuja   
27         127   23.0  Female           110.0   3/17/2022           kano   
28         128   40.0    Male           350.0   9/25/2019  Port-Harcourt   
29         129   22.0    Male           160.0   7/19/2021          Abuja   
30         130   31.0  female           280.0  11/11/2020          Lagos   
31         131   28.0    Male           270.0  13-07-2021         kaduna   
32         132   26.0  Female           150.0   12/4/2021         Kaduna   
33         133   33.0    Male             NaN   4/26/2020          Abuja   
34         134   27.0    male           190.0   2/21/2021          LAgos   
35         135   31.0  Female             NaN   5/14/2021            NaN   
36         136   29.0  Female           310.0   1/18/2022          Abuja   
37         137   28.0   FMale           200.0  10/10/2020          Abuja   
38         138   45.0    Male           400.0   9/20/2021          Lagos   
39         139   46.0  Female           420.0   3/22/2021  Port Harcourt   
40         140   34.0    Male           260.0   8/12/2019           Kano   
41         141   25.0    Male           180.0    6/1/2021         Ibadan   
42         142   29.0  Female           290.0         NaN          Abuja   
43         143   31.0    Male           150.0   7/24/2021          Lagos   
44         144   32.0  Female           230.0   5/15/2021         Kaduna   
45         145   22.0    Male            80.0   1/30/2020  Port Harcourt   
46         146  500.0  Female           300.0   8/18/2021          Lagos   
47         147   30.0    Male           200.0   9/10/2020          Abuja   
48         148   33.0  Female           210.0   2/27/2021         Ibadan   
49         149   31.0     NaN           250.0    4/6/2021          Lagos   
50         150   24.0    Male           190.0  12/30/2020          Abuja   

   IsMember  
0  

In [ ]:
y = data['Income']

In [ ]:
y.head()


0    45000.0
1    52000.0
2        NaN
3    58000.0
4    47000.0
Name: Income, dtype: float64

In [ ]:
from sklearn.impute import SimpleImputer
imputer = SimpleImputer(missing_values=np.nan, strategy='median')
imputer.fit(x[['Age','PurchaseAmount']])
x[['Age','PurchaseAmount']] = imputer.transform(x[['Age','PurchaseAmount']])


/tmp/ipython-input-2646973809.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  x[['Age','PurchaseAmount']] = imputer.transform(x[['Age','PurchaseAmount']])


In [ ]:
x

CustomerID    Age  Gender  PurchaseAmount    JoinDate       Location  \
0          101   25.0    Male           120.0   5/10/2021          Lagos   
1          102   29.0  Female           220.0   7/13/2020          Abuja   
2          103   33.0       M           310.0  13-08-2022          Lagos   
3          104  200.0  Female           220.0   9/21/2021   Port harourt   
4          105   28.0     NaN           150.0   12/2/2021         Kaduna   
5          106   22.0    Male            90.0   1/17/2020          lagos   
6          107   31.0  Female          5000.0  11/11/2019          Abuja   
7          108   29.0   Fmale           180.0   4/25/2021         Ibadan   
8          109   27.0    Male           160.0   7/18/2020         IBADAN   
9          110   35.0  Female           200.0  2022-02-30          Lagos   
10         111   26.0    male           130.0    5/5/2020           kano   
11         112   29.0  Female           300.0   8/19/2021         Kaduna   
12         113   44.0    Male           800.0   3/12/2019          Lagos   
13         114   31.0  Female           250.0   6/21/2019          Abuja   
14         115   29.0  Female           260.0   11/3/2021          Abuja   
15         115   29.0  Female           260.0   11/3/2021          Abuja   
16         116   18.0    Male            70.0   1/14/2022           Kano   
17         117   21.0     NaN         50000.0   7/22/2021          Lagos   
18         118   29.0    Male           140.0  12/18/2020  Port Harcourt   
19         119   32.0  Female           220.0    9/5/2021          Abuja   
20         120   -5.0    Male           210.0   1/25/2021          Lagos   
21         121   27.0  Female          8000.0   2/20/2019        Kadunaa   
22         122   31.0    Male           160.0    3/1/2020          Abuja   
23         123   29.0  Female           260.0   6/16/2020          Lagos   
24         124   29.0  Female           240.0  10/28/2021         Kaduna   
25         125   35.0     NaN           450.0   4/12/2021         Ibadan   
26         126   24.0  Female           100.0    8/9/2020          Abuja   
27         127   23.0  Female           110.0   3/17/2022           kano   
28         128   40.0    Male           350.0   9/25/2019  Port-Harcourt   
29         129   22.0    Male           160.0   7/19/2021          Abuja   
30         130   31.0  female           280.0  11/11/2020          Lagos   
31         131   28.0    Male           270.0  13-07-2021         kaduna   
32         132   26.0  Female           150.0   12/4/2021         Kaduna   
33         133   33.0    Male           220.0   4/26/2020          Abuja   
34         134   27.0    male           190.0   2/21/2021          LAgos   
35         135   31.0  Female           220.0   5/14/2021            NaN   
36         136   29.0  Female           310.0   1/18/2022          Abuja   
37         137   28.0   FMale           200.0  10/10/2020          Abuja   
38         138   45.0    Male           400.0   9/20/2021          Lagos   
39         139   46.0  Female           420.0   3/22/2021  Port Harcourt   
40         140   34.0    Male           260.0   8/12/2019           Kano   
41         141   25.0    Male           180.0    6/1/2021         Ibadan   
42         142   29.0  Female           290.0         NaN          Abuja   
43         143   31.0    Male           150.0   7/24/2021          Lagos   
44         144   32.0  Female           230.0   5/15/2021         Kaduna   
45         145   22.0    Male            80.0   1/30/2020  Port Harcourt   
46         146  500.0  Female           300.0   8/18/2021          Lagos   
47         147   30.0    Male           200.0   9/10/2020          Abuja   
48         148   33.0  Female           210.0   2/27/2021         Ibadan   
49         149   31.0     NaN           250.0    4/6/2021          Lagos   
50         150   24.0    Male           190.0  12/30/2020          Abuja   

   IsMember  
0       Yes  
1        No  
2       Y

# Task
Perform a hyperparameter tuning for `RandomForestRegressor` on the prepared training data (`x_train`, `y_train`) using `GridSearchCV`. Define a parameter grid that includes values for `n_estimators`, `max_depth`, `min_samples_split`, and `min_samples_leaf`. After fitting the `GridSearchCV` object, evaluate the `best_estimator_` on the test set (`x_test`, `y_test`) and print its R2 score.

## Define Parameter Grid

### Subtask:
Define a dictionary of hyperparameters and their possible values for the RandomForestRegressor to search through.


**Reasoning**:
The subtask requires defining a parameter grid for hyperparameter tuning. I will create a dictionary with the specified hyperparameters and their possible values.



In [ ]:
param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [10, 20, None],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}
print("Parameter grid defined successfully.")

Parameter grid defined successfully.


**Reasoning**:
The next logical step is to import the `GridSearchCV` module to perform hyperparameter tuning for the `RandomForestRegressor`.



In [ ]:
from sklearn.model_selection import GridSearchCV
print("GridSearchCV imported successfully.")

GridSearchCV imported successfully.


**Reasoning**:
Now that the parameter grid is defined and `GridSearchCV` is imported, I will initialize the `GridSearchCV` object with the `RandomForestRegressor` model and the defined `param_grid`. This sets up the search process for optimal hyperparameters.



In [ ]:
grid_search = GridSearchCV(estimator=RandomForestRegressor(random_state=42), param_grid=param_grid, cv=3, n_jobs=-1, verbose=2)
print("GridSearchCV initialized.")

GridSearchCV initialized.


**Reasoning**:
With the `GridSearchCV` object initialized, the next step is to fit it to the training data (`x_train`, `y_train`) to perform the hyperparameter search and find the best model.



In [ ]:
grid_search.fit(x_train, y_train)
print("GridSearchCV fitting complete.")

Fitting 3 folds for each of 81 candidates, totalling 243 fits
GridSearchCV fitting complete.


**Reasoning**:
After fitting `GridSearchCV`, I will retrieve the best estimator found during the search, which is the RandomForestRegressor model with the optimal hyperparameters.



In [ ]:
best_rf_model = grid_search.best_estimator_
print("Best RandomForestRegressor model retrieved.")

Best RandomForestRegressor model retrieved.


**Reasoning**:
Now that the best model has been retrieved, I will use it to make predictions on the test set (`x_test`) to evaluate its performance.



In [ ]:
y_pred_best_rf = best_rf_model.predict(x_test)
print("Predictions made using the best RandomForestRegressor model.")

Predictions made using the best RandomForestRegressor model.


**Reasoning**:
To evaluate the performance of the best model, I will calculate its R2 score using the actual values (`y_test`) and the predicted values (`y_pred_best_rf`).



In [ ]:
r2_best_rf = r2_score(y_test, y_pred_best_rf)
print(f"R2 score for the best RandomForestRegressor model: {r2_best_rf}")

R2 score for the best RandomForestRegressor model: 0.6515306527160604


## Summary:

### Data Analysis Key Findings

*   A parameter grid was successfully defined for `RandomForestRegressor`, including hyperparameters such as `n_estimators`, `max_depth`, `min_samples_split`, and `min_samples_leaf`.
*   `GridSearchCV` was initialized with the defined parameter grid and used to perform hyperparameter tuning across 243 fits (81 candidate combinations with 3-fold cross-validation).
*   The best `RandomForestRegressor` model, identified through `GridSearchCV`, achieved an R2 score of approximately 0.6515 on the test set.

### Insights or Next Steps

*   The current R2 score of 0.6515 suggests a moderately good fit; further investigation into feature engineering or exploring other model architectures might yield better performance.
*   Consider expanding the parameter grid or using more advanced tuning techniques like Randomized Search or Bayesian Optimization to potentially find a better performing model within a more efficient search space.
